[Cocoa] Add a way for clients to take snapshots of nodes using `_WKJSHandle`

whsieh opened the pull request:
#### d32d0c71f389cd84817e27eaf9fdab64c78d2c00
<pre>
[Cocoa] Add a way for clients to take snapshots of nodes using `_WKJSHandle`
<a href="https://bugs.webkit.org/show_bug.cgi?id=299019">https://bugs.webkit.org/show_bug.cgi?id=299019</a>
<a href="https://rdar.apple.com/160776803">rdar://160776803</a>

Reviewed by Timothy Hatcher and Abrar Rahman Protyasha.

Add support for `-[WKWebView _takeSnapshotOfNode:completionHandler:]`; see below for more details.

Test: WKWebView.SnapshotNodeByJSHandle

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _takeSnapshotOfNode:completionHandler:]):

Implement the new method on `WKWebView`. This grabs the `WebFrameProxy` using the handle&apos;s frame ID
(from frame info data), and sends an async message to take a snapshot of the node that corresponds
to the given JS handle ID.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::takeSnapshotOfNode):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::takeSnapshotOfNode):

Add plumbing from `WebFrameProxy` &lt;-&gt; `WebFrame`. We don&apos;t need to go through `WebPageProxy` in this
case, because we already have the frame identifier of the target.

* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::shareableBitmapSnapshotForNode):

Move this helper method out of iOS-specific code, and into platform-agnostic code in the `.cpp` so
we can use it in `takeSnapshotOfNode`.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::shareableBitmapSnapshotForNode): Deleted.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewSnapshot.mm:
(-[WKWebView takeSnapshotOfNode:]):
(TestWebKitAPI::TEST(WKWebView, SnapshotNodeByJSHandle)):

Add a new API test.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/node-snapshotting.html: Added.
* Tools/TestWebKitAPI/cocoa/CGImagePixelReader.cpp:
(TestWebKitAPI::CGImagePixelReader::cssColorAt const):

Add a helper method to extract the color at the given `(x, y)` coordinates, as a serialized string
for CSS.

* Tools/TestWebKitAPI/cocoa/CGImagePixelReader.h:

Canonical link: <a href="https://commits.webkit.org/300105@main">https://commits.webkit.org/300105@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/79fa85b71b935c6722caeb6401260d7ee702ac3d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121368 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41065 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31724 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127808 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73451 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9f408272-692a-4666-bc66-8ecef608280d) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123244 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41767 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49644 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92191 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/61342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c2ee336f-c17c-4ddc-9cb1-92f783ff5b27) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124320 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33353 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/108750 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72867 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3f5795b9-e138-4e09-87d9-2a50da2487cc) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32370 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26891 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71387 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102850 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27065 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130642 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48296 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36721 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100788 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48664 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104950 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100693 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25516 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46102 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24170 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44992 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48154 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53867 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47626 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50972 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49308 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewSnapshot.mm:
```diff
@@ -717,4 +752,60 @@ - (void)takeSnapshotWithWebView:(WKWebView *)webView configuration:(WKSnapshotCo
+#elif PLATFORM(IOS_FAMILY)
+    window = adoptNS([[UIWindow alloc] initWithFrame:[webView frame]]);
+    [window addSubview:webView.get()];
+#endif
```
is this necessary to test the snapshotting?

whsieh replied:
Oh, looks like it's not (I'd copied them from the above test). Will remove — thanks for the reviews!

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewSnapshot.mm:
```diff
@@ -717,4 +752,60 @@ - (void)takeSnapshotWithWebView:(WKWebView *)webView configuration:(WKSnapshotCo
+        CGImagePixelReader reader { Util::convertToCGImage(image.get()).get() };
+        EXPECT_WK_STREQ("rgb(0, 0, 0)", reader.cssColorAt(2, 2));
+        EXPECT_FALSE(reader.isTransparentBlack(reader.width() / 2, reader.height() / 2));
+    }
```
Maybe coverage for an error case? Pass in a null handle?

whsieh replied:
Yes, will add!

(I'll also add one for when the handle doesn't refer to a DOM node at all)

whsieh committed:
[Cocoa] Add a way for clients to take snapshots of nodes using `_WKJSHandle`
https://bugs.webkit.org/show_bug.cgi?id=299019
rdar://160776803

Reviewed by Timothy Hatcher and Abrar Rahman Protyasha.

Add support for `-[WKWebView _takeSnapshotOfNode:completionHandler:]`; see below for more details.

Test: WKWebView.SnapshotNodeByJSHandle

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _takeSnapshotOfNode:completionHandler:]):

Implement the new method on `WKWebView`. This grabs the `WebFrameProxy` using the handle's frame ID
(from frame info data), and sends an async message to take a snapshot of the node that corresponds
to the given JS handle ID.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::takeSnapshotOfNode):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::takeSnapshotOfNode):

Add plumbing from `WebFrameProxy` <-> `WebFrame`. We don't need to go through `WebPageProxy` in this
case, because we already have the frame identifier of the target.

* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::shareableBitmapSnapshotForNode):

Move this helper method out of iOS-specific code, and into platform-agnostic code in the `.cpp` so
we can use it in `takeSnapshotOfNode`.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::shareableBitmapSnapshotForNode): Deleted.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewSnapshot.mm:
(-[WKWebView takeSnapshotOfNode:]):
(TestWebKitAPI::TEST(WKWebView, SnapshotNodeByJSHandle)):

Add a new API test.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/node-snapshotting.html: Added.
* Tools/TestWebKitAPI/cocoa/CGImagePixelReader.cpp:
(TestWebKitAPI::CGImagePixelReader::cssColorAt const):

Add a helper method to extract the color at the given `(x, y)` coordinates, as a serialized string
for CSS.

* Tools/TestWebKitAPI/cocoa/CGImagePixelReader.h:

Canonical link: https://commits.webkit.org/300105@main
